digdag-core/src/main/java/io/digdag/core/database/DatabaseSessionStoreManager.java

```diff
@@ -1776,12 +1824,26 @@ StoredSession upsertAndLockSession(@Bind("projectId") int projectId,
                 " and s.id \\< :lastId" +
                 " and <acFilter>" +
                 " order by s.id desc" +
-                " limit :limit")
+                " limit :limit" +
+                " offset :offset")
         List<StoredSessionWithLastAttempt> getSessionsOfProject(
                 @Bind("siteId") int siteId,
                 @Bind("projId") int projId,
                 @Bind("limit") int limit,
                 @Bind("lastId") long lastId,
+                @Bind("offset") int offset,
+                @Define("acFilter") String acFilter);
+
+        @SqlQuery("select count(1)" +
```

Comment:
I'm wondering about the impact of these queries on existing endpoints. Let me check it on our environment using EXPLAIN. These only counts the number of sessions, but since these queries will execute every time a user/system accesses the `/sessions`, I _assume_ that the impact may not be small.

Reply:
If the performance impact is significant, then we will have to choose to avoid using totalCount. In this case, it will be difficult to get the number of pages, but I think there is room to consider whether this feature is necessary from the first implementation. Of course, I know it's more convenient, but I'd like to know how many use cases there are for going to the oldest sessions page. This is just my personal opinion but, I feel that it's enough to get from the newest to the oldest sessions as a first step.

Anyway, let me check these query's performance once.

Reply:
I checked the performance of the queries used in these methods.
- getSessionsCount
- getSessionsCountOfWorkflowByName
- getSessionsCountOfWorkflowByName

It seems queries used for `getSessionsCountOfWorkflowByName` and `getSessionsCountOfWorkflowByName` looks good to me. On the other hand, the query used for `getSessionsCount` seems to take a long time, I feel there is room to improve it because if there are a lot of sessions, it causes the same situation. 

### getSessionsCount

```sql
=> explain analyze 
select count(1) 
from sessions
where project_id in (select id from projects where site_id = $siteId) 
and true;
                             QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Finalize Aggregate  (cost=243313.86..243313.87 rows=1 width=8) (actual time=4480.372..4480.373 rows=1 loops=1)
   ->  Gather  (cost=243313.65..243313.86 rows=2 width=8) (actual time=4480.032..4481.470 rows=3 loops=1)
         Workers Planned: 2
         Workers Launched: 2
         ->  Partial Aggregate  (cost=242313.65..242313.66 rows=1 width=8) (actual time=4477.687..4477.688 rows=1 loops=3)
               ->  Hash Join  (cost=1535.35..240855.84 rows=583121 width=0) (actual time=14.422..3573.276 rows=3229102 loops=3)
                     Hash Cond: (sessions.project_id = projects.id)
                     ->  Parallel Index Only Scan using sessions_on_project_id_and_workflow_name_desc on sessions  (cost=0.56..228345.35 rows=4181036 width=4) (actual time=0.028..1559.563 rows=3363532 loops=3)
                           Heap Fetches: 253092
                     ->  Hash  (cost=1370.76..1370.76 rows=13122 width=4) (actual time=12.184..12.187 rows=13076 loops=3)
                           Buckets: 16384  Batches: 1  Memory Usage: 588kB
                           ->  Index Scan using projects_on_site_id_and_name on projects  (cost=0.42..1370.76 rows=13122 width=4) (actual time=0.012..7.453 rows=13076 loops=3)
                                 Index Cond: (site_id = $siteId)
 Planning Time: 0.313 ms
 Execution Time: 4481.589 ms
(15 rows)
```

### getSessionsCountOfWorkflowByName

```sql
=> explain analyze 
select count(1) 
from sessions s 
join session_attempts sa on sa.id = s.last_attempt_id 
join projects proj on proj.id = s.project_id
where s.project_id = $projectId
and sa.site_id = $siteId
and true;
                                                                                              QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=39333.09..39333.10 rows=1 width=8) (actual time=50.741..50.741 rows=1 loops=1)
   ->  Nested Loop  (cost=1001.41..39293.76 rows=15734 width=0) (actual time=0.413..47.061 rows=15630 loops=1)
         ->  Index Only Scan using projects_pkey on projects proj  (cost=0.29..1.81 rows=1 width=4) (actual time=0.022..0.024 rows=1 loops=1)
               Index Cond: (id = 43908)
               Heap Fetches: 0
         ->  Gather  (cost=1001.12..39134.61 rows=15734 width=4) (actual time=0.390..41.185 rows=15630 loops=1)
               Workers Planned: 1
               Workers Launched: 1
               ->  Nested Loop  (cost=1.12..36561.21 rows=9255 width=4) (actual time=0.038..38.218 rows=7815 loops=2)
                     ->  Parallel Index Scan using sessions_on_project_id_and_workflow_name_desc on sessions s  (cost=0.56..17363.79 rows=9641 width=12) (actual time=0.011..7.481 rows=7815 loops=2)
                           Index Cond: (project_id = 43908)
                     ->  Index Only Scan using session_attempts_on_site_id on session_attempts sa  (cost=0.56..1.99 rows=1 width=8) (actual time=0.003..0.003 rows=1 loops=15630)
                           Index Cond: ((site_id = 1) AND (id = s.last_attempt_id))
                           Heap Fetches: 25
 Planning Time: 1.072 ms
 Execution Time: 51.975 ms
(16 rows)
```

### getSessionsCountOfWorkflowByName

```sql
=> explain analyze 
select count(1)
from sessions s
join session_attempts sa on sa.id = s.last_attempt_id
join projects proj on proj.id = s.project_id
where s.project_id = $projectId
and s.workflow_name = 'test'
and sa.site_id = $siteId
and true;
           QUERY PLAN
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=25.02..25.03 rows=1 width=8) (actual time=3.665..3.666 rows=1 loops=1)
   ->  Nested Loop  (cost=1.41..25.00 rows=6 width=0) (actual time=0.073..3.568 rows=349 loops=1)
         ->  Index Only Scan using projects_pkey on projects proj  (cost=0.29..1.81 rows=1 width=4) (actual time=0.021..0.022 rows=1 loops=1)
               Index Cond: (id = 43908)
               Heap Fetches: 0
         ->  Nested Loop  (cost=1.12..23.13 rows=6 width=4) (actual time=0.050..3.392 rows=349 loops=1)
               ->  Index Scan using sessions_on_project_id_and_workflow_name_desc on sessions s  (cost=0.56..9.15 rows=6 width=12) (actual time=0.025..0.498 rows=349 loops=1)
                     Index Cond: ((project_id = $projectId) AND (workflow_name = 'test'::text))
               ->  Index Only Scan using session_attempts_on_site_id on session_attempts sa  (cost=0.56..2.33 rows=1 width=8) (actual time=0.007..0.007 rows=1 loops=349)
                     Index Cond: ((site_id = 1) AND (id = s.last_attempt_id))
                     Heap Fetches: 24
 Planning Time: 1.095 ms
 Execution Time: 3.772 ms
(13 rows)
```

Reply:
I checked the performance of the queries used in these methods.
- getSessionsCount
- getSessionsCountOfWorkflowByName
- getSessionsCountOfWorkflowByName

It seems queries used for `getSessionsCountOfWorkflowByName` and `getSessionsCountOfWorkflowByName` looks good to me. On the other hand, the query used for `getSessionsCount` seems to take a long time, I feel there is room to improve it because if there are a lot of sessions, it causes the same situation. 

### getSessionsCount

```sql
=> explain analyze 
select count(1) 
from sessions
where project_id in (select id from projects where site_id = $siteId) 
and true;
                             QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Finalize Aggregate  (cost=243313.86..243313.87 rows=1 width=8) (actual time=4480.372..4480.373 rows=1 loops=1)
   ->  Gather  (cost=243313.65..243313.86 rows=2 width=8) (actual time=4480.032..4481.470 rows=3 loops=1)
         Workers Planned: 2
         Workers Launched: 2
         ->  Partial Aggregate  (cost=242313.65..242313.66 rows=1 width=8) (actual time=4477.687..4477.688 rows=1 loops=3)
               ->  Hash Join  (cost=1535.35..240855.84 rows=583121 width=0) (actual time=14.422..3573.276 rows=3229102 loops=3)
                     Hash Cond: (sessions.project_id = projects.id)
                     ->  Parallel Index Only Scan using sessions_on_project_id_and_workflow_name_desc on sessions  (cost=0.56..228345.35 rows=4181036 width=4) (actual time=0.028..1559.563 rows=3363532 loops=3)
                           Heap Fetches: 253092
                     ->  Hash  (cost=1370.76..1370.76 rows=13122 width=4) (actual time=12.184..12.187 rows=13076 loops=3)
                           Buckets: 16384  Batches: 1  Memory Usage: 588kB
                           ->  Index Scan using projects_on_site_id_and_name on projects  (cost=0.42..1370.76 rows=13122 width=4) (actual time=0.012..7.453 rows=13076 loops=3)
                                 Index Cond: (site_id = $siteId)
 Planning Time: 0.313 ms
 Execution Time: 4481.589 ms
(15 rows)
```

### getSessionsCountOfWorkflowByName

```sql
=> explain analyze 
select count(1) 
from sessions s 
join session_attempts sa on sa.id = s.last_attempt_id 
join projects proj on proj.id = s.project_id
where s.project_id = $projectId
and sa.site_id = $siteId
and true;
                                                                                              QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=39333.09..39333.10 rows=1 width=8) (actual time=50.741..50.741 rows=1 loops=1)
   ->  Nested Loop  (cost=1001.41..39293.76 rows=15734 width=0) (actual time=0.413..47.061 rows=15630 loops=1)
         ->  Index Only Scan using projects_pkey on projects proj  (cost=0.29..1.81 rows=1 width=4) (actual time=0.022..0.024 rows=1 loops=1)
               Index Cond: (id = 43908)
               Heap Fetches: 0
         ->  Gather  (cost=1001.12..39134.61 rows=15734 width=4) (actual time=0.390..41.185 rows=15630 loops=1)
               Workers Planned: 1
               Workers Launched: 1
               ->  Nested Loop  (cost=1.12..36561.21 rows=9255 width=4) (actual time=0.038..38.218 rows=7815 loops=2)
                     ->  Parallel Index Scan using sessions_on_project_id_and_workflow_name_desc on sessions s  (cost=0.56..17363.79 rows=9641 width=12) (actual time=0.011..7.481 rows=7815 loops=2)
                           Index Cond: (project_id = $projectId)
                     ->  Index Only Scan using session_attempts_on_site_id on session_attempts sa  (cost=0.56..1.99 rows=1 width=8) (actual time=0.003..0.003 rows=1 loops=15630)
                           Index Cond: ((site_id = 1) AND (id = s.last_attempt_id))
                           Heap Fetches: 25
 Planning Time: 1.072 ms
 Execution Time: 51.975 ms
(16 rows)
```

### getSessionsCountOfWorkflowByName

```sql
=> explain analyze 
select count(1)
from sessions s
join session_attempts sa on sa.id = s.last_attempt_id
join projects proj on proj.id = s.project_id
where s.project_id = $projectId
and s.workflow_name = 'test'
and sa.site_id = $siteId
and true;
           QUERY PLAN
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=25.02..25.03 rows=1 width=8) (actual time=3.665..3.666 rows=1 loops=1)
   ->  Nested Loop  (cost=1.41..25.00 rows=6 width=0) (actual time=0.073..3.568 rows=349 loops=1)
         ->  Index Only Scan using projects_pkey on projects proj  (cost=0.29..1.81 rows=1 width=4) (actual time=0.021..0.022 rows=1 loops=1)
               Index Cond: (id = 43908)
               Heap Fetches: 0
         ->  Nested Loop  (cost=1.12..23.13 rows=6 width=4) (actual time=0.050..3.392 rows=349 loops=1)
               ->  Index Scan using sessions_on_project_id_and_workflow_name_desc on sessions s  (cost=0.56..9.15 rows=6 width=12) (actual time=0.025..0.498 rows=349 loops=1)
                     Index Cond: ((project_id = $projectId) AND (workflow_name = 'test'::text))
               ->  Index Only Scan using session_attempts_on_site_id on session_attempts sa  (cost=0.56..2.33 rows=1 width=8) (actual time=0.007..0.007 rows=1 loops=349)
                     Index Cond: ((site_id = 1) AND (id = s.last_attempt_id))
                     Heap Fetches: 24
 Planning Time: 1.095 ms
 Execution Time: 3.772 ms
(13 rows)
```

Reply:
@szyn 
Thank you for your performance check.
So we should consider to resort to one of following approaches.
- Add an index to the database to improve the performance
- Change the implementation to improve the performance
- Employ a more light weight UI (e.g. putting only "Latest" button and  "Older" button rather than page number buttons)

I think I can solely implement the last approach and might need your help or take a long time to implement first and second approach.

Do you have any opinions?

Reply:
If I understand correctly, the last approach is what you pointed before check the performance. https://github.com/treasure-data/digdag/pull/1515#discussion_r638664719
> This is just my personal opinion but, I feel that it's enough to get from the newest to the oldest sessions as a first step.

Reply:
Thank you for checking. I was thinking the same things as you. From my point of view, as you pointed out, the first one/second one will take a long time. How about we proceed with the last approach at first? Of course, I appreciate doing 1/2 as the next improvement because this PR's (https://github.com/treasure-data/digdag/pull/1506) pagination approach that you wrote was very good to me.

Reply:
Thank you for checking. I was thinking the same things as you. From my point of view, as you pointed out, the first one/second one will take a long time. How about we proceed with the last approach at first? Of course, I appreciate doing the first/second one as the next improvement because this PR's (https://github.com/treasure-data/digdag/pull/1506) pagination approach that you write is very good to me.

Reply:
Thank you for your response.
Then I'll implement "Latest" button and "Older" (might be "Previous", "Earlier") button UI.

I close this PR for a time.
I will not work on the first / second ones for a long time.
Because I think the last one is enough for most cases and I prioritize other issues even I contribute this project.